Sources/SwiftFusion/Inference/ChordalInitialization.swift

```diff
@@ -0,0 +1,205 @@
+      let UVT: Tensor<Double> = matmul(U!, V!.transposed())
+      
+      let det = (UVT[0, 0].scalar! * (UVT[1, 1].scalar! * UVT[2, 2].scalar! - UVT[2, 1].scalar! * UVT[1, 2].scalar!)
+                 - UVT[1, 0].scalar! * (UVT[0, 1].scalar! * UVT[2, 2].scalar! - UVT[2, 1].scalar! * UVT[0, 2].scalar!)
+                 + UVT[2, 0].scalar! * (UVT[0, 1].scalar! * UVT[1, 2].scalar! - UVT[1, 1].scalar! * UVT[0, 2].scalar!))
+      
+      let R = matmul(matmul(U!, Tensor<Double>(shape: [3], scalars: [1, 1, det]).diagonal()), V!.transposed()).transposed()
+      
+      let initRot = Rot3(coordinate:
+                          Matrix3Coordinate(Matrix3(R[0, 0].scalar!, R[0, 1].scalar!, R[0, 2].scalar!,
+                                                    R[1, 0].scalar!, R[1, 1].scalar!, R[1, 2].scalar!,
+                                                    R[2, 0].scalar!, R[2, 1].scalar!, R[2, 2].scalar!))
+      )
+      
+      // TODO(fan): relies on the assumption of continous and ordered allocation
+      let _ = validRot3.store(initRot)
+    }
+    return validRot3;
+  }
+  
+  /// This function computes the rotations
+  public func computeOrientationsChordal(graph: FactorGraph, val: VariableAssignments, ids: Array<TypedID<Pose3, Int>>) -> VariableAssignments {
+    var relaxedRot3 = VariableAssignments()
+    var associations = Dictionary<Int, TypedID<Vector9, Int>>()
+    
+    for v in ids {
+      let R = val[v].rot.coordinate.R
+      associations[v.perTypeID] = relaxedRot3.store(
+        Vector9(R.s00, R.s01, R.s02,
+        R.s10, R.s11, R.s12,
+        R.s20, R.s21, R.s22))
+    }
+    
+    let R_a = val[anchorId].rot.coordinate.R
+    associations[anchorId.perTypeID] = relaxedRot3.store(Vector9(R_a.s00, R_a.s01, R_a.s02,
+    R_a.s10, R_a.s11, R_a.s12,
+    R_a.s20, R_a.s21, R_a.s22))
+    
+    // regularize measurements and plug everything in a factor graph
+    let relaxedGraph: GaussianFactorGraph = buildLinearOrientationGraph(g: graph, v: val, v_linear: relaxedRot3, associations: associations)
+
+    // Solve the LFG
+    var optimizer = GenericCGLS()
+    print("[COC BEFORE] \(relaxedGraph.errorVectors(at: relaxedRot3).squaredNorm)")
+    optimizer.optimize(gfg: relaxedGraph, initial: &relaxedRot3)
+    print("[COC AFTER ] \(relaxedGraph.errorVectors(at: relaxedRot3).squaredNorm)")
+    
+    for i in associations.sorted(by: { $0.0 < $1.0 }) {
+      print("\(i.key): \(relaxedRot3[i.value])")
+    }
+    
+    print("\(ids)")
+    let result = normalizeRelaxedRotations(relaxedRot3, associations: associations, ids: ids)
+    
+    // normalize and compute Rot3
+    return result
+  }
+  
+  /// This function computes the inital poses given the chordal initialized rotations.
+  public func computePoses(graph: FactorGraph, original_initialization: VariableAssignments, orientations: VariableAssignments, ids: Array<TypedID<Pose3, Int>>) -> VariableAssignments {
```

Comment:
original_initialization seems to suggest that this function is not independent. Think about inputs and outputs of the computation below.

Reply:
This is to work around that fact that we are operating on a subgraph, so we need this to map the subgraph variables to the original graph.

Reply:
I understand (I think) why you're doing it. I urge you to think about better design

Reply:
It's kind of not avoidable given the current `TypedID` design, as `TypedID`s can only be allocated in a consecutive manner ;)

Reply:
Think harder

Reply:
done